ls1intum/Artemis
# `Development`: Add ability to disable quiz subsystem

### Checklist
#### General


#### Server
#### Client

### Motivation and Context

This PR is based on #5790.
We also want to decouple the quiz subsystem so that not all Artemis instances provide the quiz-logic.

### Description
We implement the following changes:
- Block access to quiz exercises if not available on the system
- Only handle Quiz logic by instances that run the quiz profile
  - REST-calls are routed by nginx to the correct instances
  - WebSocket messages are either handled directly if the user is connected to a instance that runs quiz logic or forwarded to another instance using JMS-messages.
    - The JMS messages are distributed by the broker (ActiveMQ in our case) in a queue-logic so that the requests are only handled once.
    - One instance will process the message and act accordingly (by e.g. saving the submission to the database).
    - Other instances will not handle the message

### Steps for Testing
Testing this PR is not possible on the normal test servers.
On normal test servers you can only ensure that all quiz-related functionality is still working.


If you have a multi-node setup (including a broker!), you can test that quiz exercises are only saved by one of the instances that run the quiz profile.


#### Exam Mode Testing
- Make sure that the quiz still works as expected in the exam mode when the functionality is available (esp. on the staging environment)
- On an instance with disabled functionality, quizzes should not be added to exams.

### Review Progress






#### Code Review
#### Manual Tests
#### Exam Mode Test

### Test Coverage





### Screenshots


Buttons disabled for students:
![image](https://github.com/ls1intum/Artemis/assets/5084100/51f033ca-9a46-44a3-b839-5f532e4681b1)

Instructors are unable to create quiz exercises:
![image](https://github.com/ls1intum/Artemis/assets/5084100/f24b51c0-0b76-4cba-9f4a-294f351d3960)


Filtering quiz exercises is not possible:
![image](https://github.com/ls1intum/Artemis/assets/5084100/154db74e-8618-46a1-85f6-2c4768bb0d5e)
